Feature zero token nodes

aleksbykov opened the pull request:
New feature Zero token nodes - were presented by https://github.com/scylladb/scylladb/pull/19684. These nodes are normal scylla node but they don't joined to token ring of  cluster and don't contain any user data. They are used by raft and group0 to preserve majority for raft operations
Zero token nodes could be created only upon bootstrap if scylla.yaml contains `join_ring: false`. 
Zero token nodes are normal scylla nodes and all operations are allowed with them (remove, decommission, replace with zero token nodes only)
 the main goal of them:
 - preserve quorum and majority in multi dc config. ex dc1: 3 : dc2: 3 - if one dc die or dc become isolated, both dc will lost majority and no any topology or schema operation will be allowed, if will be add one more dc with 1 zero token node dc3: 1, then majority will be saved and 
 - because zero token nodes stay as group0 voters, even single dc cluster: 4 nodes, +1 zero token nodes could save cluster majority and even if 2 nodes are die,  schema changes and topology operations will be allowed.

zero token node doesn't contain user data, therefore the instance type could be smaller
Drivers don't  see zerotoken nodes and ignore them.

To add support zero token nodes in sct:

- [x] Add special flag to Node class which will true if node is zero token
- [x] Add new sct_config parameter to 
      - number of zero token nodes
      - add additional dc with zero token nodes only
- [ ] init cluster with zero token nodes:
   here we have 2 ways:
   1. - [ ]  using new parameter, mark latest `num_zero_token` in list as zero token nodes
   This solution has several benefits:
      - don't need to fix provisioner code, all backends will be supported (aws, gce, docker, k8s??)
      - if num_zero_tokens = 0, nothing will be changed
      - simple to implement
      - just add new test with correct yaml configuration
    and disadvantages:
     - zero token nodes will have same instance type as normal db nodes - in case test will not have a lot of them, it will not be a problem
     - if enable by default even 1 zero token node in config, all config yaml files should be reviewed so number of normal nodes was relevant to RF of schemas, because if test is conifgured with 3 nodes and one of them 1 zero token node and rf =3, after provision, cluster will have 2 normal nodes and 1 zero token nodes. This could affect on c-s and other stress tools.
   2. [x] - Add code to Provisioner (aws, azure) and to gce, docker, k8s?? and provision correct number of nodes n_db_nodes + nemesis_add_nodes+num_zero_token_nodes. Also add new sct config parameter which allow to set small instance type for zero token nodes
   this solution has next benefits:
    - instance type could be set independently for normal nodes and zero token nodes.
    - No need to review and fix all test config yamls 
   Disadvantages:
    - quite complex code refactor: need to fix Provisioner and all related stuff, add new code for gce, docker (we don't have provisioner for them yet, am i right?)
    - Add additional checks for instance type, so not set small instance as normal node
    - others, not yet found them )

- [x] Add support  for cluster operation and Disrupt Nemesis:
    - bootstrap new node as zero token
    - if  decommission zero token node and new zero token nodes
    - replace zero token node with zero token node only
    - remove zero token node and if after that add new node it should be zero token node
- [x] Cluster status and health
   Currently node tool status doesn't display zero-token node https://github.com/scylladb/scylladb/issues/19849. Once it will be fixed, all zero token nodes will be displayed in output as usual nodes. But before that, it is need work around so cluster health and nodetool status correctly work and don't fail the test.

- [ ] New nemesis:
    - [ ] Create new DC with zero token node only (could be improved current one)
    - [ ] Multi DC (2 dc) create new DC with zero-token node only and stop all nodes in 1 dc, check that majority is not lost ( another variations is 3 dc initiated, and nemesis check that we have 3 DCes and 1 dc has only zerotoken node and only then run, otherwise it skipped)
 

### Testing
The pr is already contains first way cluster initialization. Job done. Node with zero token is added. But due to initializtion https://github.com/scylladb/scylladb/issues/19849, cluster health is failed and job marked as failed.
- [Implementation 1 ](https://jenkins.scylladb.com/view/staging/job/scylla-staging/job/abykov/job/longevity-100gb-4h-zero-token-nodes/5)

### PR pre-checks (self review)
<!--- PR should be created as Draft, when CI finished and relevant checkboxes selected, add reviewers and then click on "Ready for review" button.-->
<!--- Put an `x` in all the boxes that apply or create PR and then click on all relevant checkboxes: -->
- [x] I added the relevant `backport` labels
- [x] I didn't leave commented-out/debugging code

### Reminders

- Add New configuration option and document them (in `sdcm/sct_config.py`)
- Add unit tests to cover my changes (under `unit-test/` folder)
- Update the Readme/doc folder relevant to this change (if needed)

aleksbykov commented:
@soyacz @fruch @roydahan @vponomaryov  @temichus need your opinion for implementation.

soyacz commented on sdcm/cluster.py:
```diff
@@ -3913,11 +3916,19 @@ def critical_node_setup_events():
             # setup in parallel
-            for node in node_list:
+
+            for node in node_list[:-num_of_zero_token_nodes]:
```
growing cluster by 1 will break the test.

soyacz commented on sdcm/provision/scylla_yaml/scylla_yaml.py:
```diff
@@ -352,6 +352,7 @@ def set_authorizer(cls, authorizer: str):
     force_gossip_topology_changes: bool = None  # False
 
     reader_concurrency_semaphore_cpu_concurrency: int = None
+    join_ring: bool = None,  # False is default
```
still SCT must be aware of zero token nodes, yaml param is not enough. See my proposal to have them as separate pool instead of adding them to `cluster.nodes`

roydahan replied:
I agree with @soyacz idea of separation.

aleksbykov replied:
Implementation is next:
 - nodes - pool of all nodes in sct
 - data_nodes - property with only data_nodes
 - zero_nodes - property with only znodes
Nemesis use target only from data_node pool
I go over all places and switch to use correct pool of nodes

soyacz commented on sdcm/sct_config.py:
```diff
@@ -1619,6 +1619,12 @@ class SCTConfiguration(dict):
+        dict(name="num_zero_token_nodes", env="SCT_NUM_ZERO_TOKEN_NODES", type=int,
+             help="""Number of zero token nodes"""),
+
+        dict(name="arbitor_dc", env="SCT_ARBITOR_DC", type=int,
```
As I proposed, zero token nodes configuration should be similar to n_db_nodes, so we could define number of them per dc.

roydahan replied:
Exactly.

soyacz reviewed (commented):
> To add support zero token nodes in sct:
> 
> * [ ]  Add special flag to Node class which will true if node is zero token
> * [ ]  Add new sct_config parameter to
>   - number of zero token nodes
>   - add additional dc with zero token nodes only
> * [ ]  init cluster with zero token nodes:
>   here we have 2 ways:
>   
>   1. * [ ]   using new parameter, mark latest `num_zero_token` in list as zero token nodes
>        This solution has several benefits:
>      * don't need to fix provisioner code, all backends will be supported (aws, gce, docker, k8s??)
>      * if num_zero_tokens = 0, nothing will be changed
>      * simple to implement
>      * just add new test with correct yaml configuration

Still, many nemesis should not take zero token nodes as the target (e.g. compaction, refresh, even enospc is rather not relevant and many more).
We could overcome it by adding another flag to specific nemeses. 

Additionally we should not be adding them to `cluster.nodes`, instead we could create another attribute `cluster.zero_token_nodes` - being in a separate pool so it's not affecting all the places we base on number of nodes.
Also should get different name, so we properly sort them out in sct test step in Jenkins (after separate provision step).

In my opinion, zero token nodes also should have own set of nemesis, specifically targeted into them, so we would see in Argus nemesis list which were run against them.

>        and disadvantages:
>   
>   
>   * zero token nodes will have same instance type as normal db nodes - in case test will not have a lot of them, it will not be a problem

We can override instance type for them, not a problem (does not require special provisioner).

>   * if enable by default even 1 zero token node in config, all config yaml files should be reviewed so number of normal nodes was relevant to RF of schemas, because if test is conifgured with 3 nodes and one of them 1 zero token node and rf =3, after provision, cluster will have 2 normal nodes and 1 zero token nodes. This could affect on c-s and other stress tools.
>   

could overcome it if sct_config code added `num_zero_token` to number of db nodes. Also I think should be renamed to `n_db_zero_token_nodes` and understood similarly to `n_db_nodes`. 

>   
>   2. [ ]  - Add code to Provisioner (aws, azure) and to gce, docker, k8s?? and provision correct number of nodes n_db_nodes + nemesis_add_nodes+num_zero_token_nodes. Also add new sct config parameter which allow to set small instance type for zero token nodes
>      this solution has next benefits:
>   
>   
>   * instance type could be set independently for normal nodes and zero token nodes.
>   * No need to review and fix all test config yamls
>     Disadvantages:
>   * quite complex code refactor: need to fix Provisioner and all related stuff, add new code for gce, docker (we don't have provisioner for them yet, am i right?)

I don't see why it requires fix in provisioner.


>   * Add additional checks for instance type, so not set small instance as normal node
>   * others, not yet found them )
> * [ ]  Add support  for cluster operation and Disrupt Nemesis:
>   
>   * bootstrap new node as zero token
>   * if  decommission zero token node and new zero token nodes
>   * replace zero token node with zero token node only
>   * remove zero token node and if after that add new node it should be zero token node
> * [ ]  Cluster status and health
>   Currently node tool status doesn't display zero-token node [nodetool status should show zero-token nodes scylladb#19849](https://github.com/scylladb/scylladb/issues/19849). Once it will be fixed, all zero token nodes will be displayed in output as usual nodes. But before that, it is need work around so cluster health and nodetool status correctly work and don't fail the test.
> * [ ]  New nemesis:
>   
>   * [ ]  Create new DC with zero token node only (could be improved current one)
>   * [ ]  Multi DC (2 dc) create new DC with zero-token node only and stop all nodes in 1 dc, check that majority is not lost ( another variations is 3 dc initiated, and nemesis check that we have 3 DCes and 1 dc has only zerotoken node and only then run, otherwise it skipped)

roydahan commented on test-cases/longevity/longevity-100gb-4h.yaml:
```diff
@@ -6,15 +6,19 @@ stress_cmd: ["cassandra-stress write cl=QUORUM duration=240m -schema 'replicatio
-n_db_nodes: 6
-n_loaders: 2
+n_db_nodes: 5
+n_db_zero_token_nodes: 1
```
What is the difference between this `n_db_zero_token_nodes` and `num_zero_token_nodes`?

aleksbykov replied:
left only n_db_zero_token_nodes

roydahan commented on defaults/test_default.yaml:
```diff
@@ -266,3 +266,6 @@ run_scylla_doctor: false
 enterprise_disable_kms: false
+
+num_zero_token_nodes: 0
+arbitor_dc: 0
```
I thought they got rid of this concept of arbitor_dc.

aleksbykov replied:
removed obsolete config param

roydahan commented on test-cases/longevity/longevity-100gb-4h.yaml:
```diff
@@ -6,15 +6,19 @@ stress_cmd: ["cassandra-stress write cl=QUORUM duration=240m -schema 'replicatio
 n_monitor_nodes: 1
 seeds_num: 3
+num_zero_token_nodes: 1
+zero_token_instance_type_db: 'i4i.xlarge'
```
I suggest to make it something tiny, as any user would do.
it shouldn't even be i4i, could be something that doesn't have NVMe or has very small NVMe.

It probably raises a requirement for scylla-machine-image to support such instance type for zero-token-node (shouldn't block the installation).

fruch commented on sdcm/nemesis.py:
```diff
@@ -1298,7 +1302,9 @@ def _add_and_init_new_cluster_nodes(self, count, timeout=MAX_TIME_WAIT_FOR_NEW_N
-        self.cluster.wait_for_nodes_up_and_normal(nodes=new_nodes)
+        # TODO: rewrite up and normal nodes
+        # skip all zero toke nodes for now
+        self.cluster.wait_for_nodes_up_and_normal(nodes=[node for node in new_nodes if not node._zero_token_node])
```
those nodes are not shown on `nodetool status`  ? we are we waiting for them ?

if we shouldn't be waiting for any reason, we might want an argument in this function, skipping, instead of logic like that

kbr-scylla replied:
> those nodes are not shown on nodetool status ? we are we waiting for them ?

this is WIP: https://github.com/scylladb/scylladb/pull/20909

aleksbykov replied:
Added workaround to use cql request

aleksbykov commented:
New changes are ready:
 - Removed unused sct_config variables. Now only 2 variables are used: n_db_zero_token_nodes and zero_token_instance_type_db. n_db_zero_token_nodes has same format as n_db_nodes: int or "0 0 1" which means that we should have 3 dc, and znode will be only in 3rd one.  zero_token_instance_type_db - instance type of znodes. if it is not set, then db_instance_type will be used.
 - Add to sct aws provisioner functionality to create instances for znodes with specified instnace type. znode instance will have tag ZeroTokenNode=True, which to detect the znode by sct and correctly configure it in scylla.yaml
 - sct (aws part only) - can detect znode and mark them with new property BaseNode._zero_token_node. Also we can create znode with aws cluster(added appropriate parameter to add_nodes() method
 - BaseScyllaCluster has 2 new properties: data_node and zero_nodes. Each return list for relevant nodes. BaseScyllaClluster.nodes was not renamed. Because it require a lot of changes in sct code. BaseScyllaClluster.nodes contains full list of nodes(datanodes and znodes).  This is also done, because znode are regular scylla node(without data) and most of the operations could be start from them
 - Nemsis.get_target_node switched for now to get target node only from datanode pool. 
 - Update all Nemesis and sct operation to use specified pool of nodes is in progress
 - Added 2 new nemesises:
   - Bootstrap/Decommission znode to cluster (singleDC and MultiDC)
   - Add new remove new dc with znode only (singleDC), For multiDC is in progress
 - Add workaround for health_validator and wait_node_up to get status of nodes from system.cluster_status table instead of `nodetool status`
- Added 2 new jobs single dc and multidc with znode configured.
   - both job have passed with regular nemesis

fruch commented:
I would recommend adding issues for follow on other backends, we can help with docker/gce/azure, and cloud/operator team with k8s backend, they would need it as well at some point.

kbr-scylla commented on sdcm/sct_config.py:
```diff
@@ -2075,10 +2082,41 @@ def __init__(self):  # noqa: PLR0912, PLR0914, PLR0915
         # 20 Validate Manager agent backup general parameters
         if backup_params := self.get("mgmt_agent_backup_config"):
             self["mgmt_agent_backup_config"] = AgentBackupParameters(**backup_params)
+        # 22 Validate zero token nodes
```
by the way, explicit numbering of steps in comments is not a good idea.

aleksbykov replied:
fixed

soyacz commented on sdcm/nemesis.py:
```diff
@@ -5147,6 +5151,53 @@ def disrupt_disable_binary_gossip_execute_major_compaction(self):
+            status = self.tester.db_cluster.get_cluster_status()
+            new_dc_list = [dc for dc in list(status.keys()) if dc.endswith("_nemesis_dc")]
+            assert new_dc_list, "new datacenter was not registered"
+            InfoEvent(message='Running full cluster repair on each node').publish()
+            for cluster_node in self.cluster.data_nodes:
+                cluster_node.run_nodetool(sub_cmd="repair -pr", publish_event=True)
```
AFAIR this step is required for adding new DC per our docs. But I don't think it should be required for znodes.

soyacz commented on sdcm/nemesis.py:
```diff
@@ -5147,6 +5151,53 @@ def disrupt_disable_binary_gossip_execute_major_compaction(self):
+            node_added = True
+            status = self.tester.db_cluster.get_cluster_status()
+            new_dc_list = [dc for dc in list(status.keys()) if dc.endswith("_nemesis_dc")]
+            assert new_dc_list, "new datacenter was not registered"
```
is this the only thing to assert here? shouldn't we verify this node got some raft status or something?

aleksbykov replied:
removed

soyacz commented on sdcm/nemesis.py:
```diff
@@ -6660,3 +6711,21 @@ class EndOfQuotaNemesis(Nemesis):
+
+class GrowShrinkZeroTokenNode(Nemesis):
+
+    disruptive = True
```
I think we used this flag to exclude 2 parallel disruptive nemesis to avoid issues with CL=QUORUM, but here I don't consider this as an issue. Maybe these should not be marked as disruptive?

aleksbykov replied:
Present new flag

soyacz commented on sdcm/nemesis.py:
```diff
@@ -6660,3 +6711,21 @@ class EndOfQuotaNemesis(Nemesis):
         self.disrupt_end_of_quota_nemesis()
+
+
+class GrowShrinkZeroTokenNode(Nemesis):
```
I think we need another nemesis flag like 'zero_token' to mark these nemesis that work with this feature, so we could exclude/include them in tests easily. @roydahan WDYT?

aleksbykov replied:
added

soyacz reviewed (commented):
Still need to verify places where `self.nodes` that include reference to all nodes (including zero token ones) is not relevant.

Besides, I wonder if we shouldn't add a test config flag that we want to use zero token nodes in given test and set it to false on all backends and enable it to some tests only.
This way we could remove all the places we verify backend and have control whether we want to test in given job at all. @roydahan WDYT?

aleksbykov commented:
Fixed according comments. New:
 - Added new global flag 'use_zero_nodes' which allow to enabled/disable support of zero nodes in sct. If it is disabled, sct will configure and use only data nodes, all nemesis related to zero nodes only will be skipped, all nemesis which could work with any nodes will work only with data_nodes.
 - Nemesis Class has new attribute 'target_node_pool' and method set_target_node_pool. This allow to explicitly configure the pool of nodes for choosing target node for nemesis. Several nemesis marked with decorator, which explicitly set  target_node_pool. by default data_nodes set as target_node_pool
 - Mark several disrupt_* method to use appropriate pool
 - Go through sct code, and set to use data_nodes instead of all nodes (self.nodes).
 - Add appropirate flag to zero node related yaml and jobs.
 - 2 testing jobs are added and will be used for testing zero nodes, all other nodes will continue to work without zero nodes for a while.
 - Added new nemesis flag 'zero_node_changes' which allow to select nemesis  which could work with zero nodes.

aleksbykov commented on sdcm/cluster.py:
```diff
@@ -306,6 +306,7 @@ def __init__(self, name, parent_cluster, ssh_login_info=None, base_logdir=None,
         self.scylla_network_configuration = None
         self._datacenter_name = None
         self._node_rack = None
+        self._zero_token_node = False
```
renamed

aleksbykov commented on sdcm/cluster_aws.py:
```diff
@@ -132,7 +132,7 @@ def instance_profile_name(self) -> str | None:
         return None
 
-    def _create_on_demand_instances(self, count, interfaces, ec2_user_data, dc_idx=0, instance_type=None):  # pylint: disable=too-many-arguments
+    def _create_on_demand_instances(self, count, interfaces, ec2_user_data, dc_idx=0, instance_type=None, zero_token=False):  # pylint: disable=too-many-arguments
```
renamed

aleksbykov commented on sdcm/provision/aws/capacity_reservation.py:
```diff
@@ -35,7 +35,7 @@ def _get_cr_request_based_on_sct_config(params) -> Tuple[dict[str, int], int]:
         instance_counts = defaultdict(int)
         nemesis_node_count = params.get("nemesis_add_node_cnt") or 0
 
-        cluster_max_size = (params.get("cluster_target_size") or params.get("n_db_nodes"))
```
fixed

aleksbykov commented on sdcm/sct_config.py:
```diff
@@ -2075,10 +2082,41 @@ def __init__(self):  # noqa: PLR0912, PLR0914, PLR0915
             self["mgmt_agent_backup_config"] = AgentBackupParameters(**backup_params)
+        # 22 Validate zero token nodes
+        zerotoken_nodes_num = self.get("n_db_zero_token_nodes")
+        token_nodes_num = self.get("n_db_nodes")
```
fixed

aleksbykov commented on sdcm/sct_config.py:
```diff
@@ -2075,10 +2082,41 @@ def __init__(self):  # noqa: PLR0912, PLR0914, PLR0915
+        # 22 Validate zero token nodes
+        zerotoken_nodes_num = self.get("n_db_zero_token_nodes")
+        token_nodes_num = self.get("n_db_nodes")
+        if backend == "aws" and zerotoken_nodes_num:
```
done

aleksbykov commented on sdcm/sct_config.py:
```diff
@@ -2075,10 +2082,41 @@ def __init__(self):  # noqa: PLR0912, PLR0914, PLR0915
+        total_nodes = token_nodes_num[:]
+
+        # temp for aws, but should work for any backend
+        if self.get('cluster_backend') == "aws" and zerotoken_nodes_num:
```
changed

aleksbykov commented on sdcm/utils/aws_utils.py:
```diff
@@ -253,9 +253,11 @@ def init_monitoring_info_from_params(monitor_info: dict, params: dict, regions:
 def init_db_info_from_params(db_info: dict, params: dict, regions: List, root_device: str = None):
     if db_info['n_nodes'] is None:
-        n_db_nodes = params.get('n_db_nodes')
+        n_db_nodes = params.total_n_db_nodes
```
fixed

aleksbykov commented on sdcm/cluster.py:
```diff
@@ -4461,11 +4520,14 @@ def get_node_status_dictionary(self, ip_address=None, verification_node=None):
                   message="Waiting for nodes to join the cluster", timeout=timeout)
         def _wait_for_nodes_up_and_normal():
-            self.check_nodes_up_and_normal(nodes=nodes, verification_node=verification_node)
+            if use_cql:
```
removed

aleksbykov commented on sdcm/nemesis.py:
```diff
@@ -1274,14 +1274,16 @@ def _replace_cluster_node(self, old_node_ip=None, host_id=None,
         return new_node
 
-    def _add_and_init_new_cluster_nodes(self, count, timeout=MAX_TIME_WAIT_FOR_NEW_NODE_UP, rack=None, instance_type: str = None) -> list[BaseNode]:
+    def _add_and_init_new_cluster_nodes(self, count, timeout=MAX_TIME_WAIT_FOR_NEW_NODE_UP, rack=None, instance_type: str = None, zero_token: bool = False) -> list[BaseNode]:
```
done

aleksbykov commented on sdcm/nemesis.py:
```diff
@@ -5147,6 +5151,53 @@ def disrupt_disable_binary_gossip_execute_major_compaction(self):
+            raise UnsupportedNemesis("Operator doesn't support multi-DC yet. Skipping.")
+        if self.cluster.test_config.MULTI_REGION:
+            raise UnsupportedNemesis(
+                "add_remove_dc skipped for multi-dc scenario (https://github.com/scylladb/scylla-cluster-tests/issues/5369)")
```
done

aleksbykov commented on sdcm/nemesis.py:
```diff
@@ -4575,9 +4575,13 @@ def _verify_cdc_feature_status(self, keyspace: str, table: str, cdc_settings: di
-        new_node = skip_on_capacity_issues(self.cluster.add_nodes)(
-            1, dc_idx=0, enable_auto_bootstrap=True)[0]  # add node
+    def _add_new_node_in_new_dc(self, zero_token=False) -> BaseNode:
+        if self.tester.params.get("cluster_backend") == "aws":
```
done

aleksbykov commented on sdcm/nemesis.py:
```diff
@@ -5147,6 +5151,53 @@ def disrupt_disable_binary_gossip_execute_major_compaction(self):
             self.target_node.restart_scylla_server()
             raise
 
+    def disrupt_grow_shrink_znodes(self):
```
done

aleksbykov commented on sdcm/nemesis.py:
```diff
@@ -5147,6 +5151,53 @@ def disrupt_disable_binary_gossip_execute_major_compaction(self):
+        system_keyspaces = ["system_distributed", "system_traces"]
+        if not node.raft.is_consistent_topology_changes_enabled:  # auth-v2 is used when consistent topology is enabled
+            system_keyspaces.insert(0, "system_auth")
+        self._switch_to_network_replication_strategy(self.cluster.get_test_keyspaces() + system_keyspaces)
```
Removed for a while

aleksbykov commented on sdcm/sct_config.py:
```diff
@@ -2075,10 +2086,39 @@ def __init__(self):  # noqa: PLR0912, PLR0914, PLR0915
     def log_config(self):
         self.log.info(self.dump_config())
 
+    @property
```
not need for now

aleksbykov commented on sdcm/sct_config.py:
```diff
@@ -2487,6 +2530,10 @@ def _check_partition_range_with_data_validation_correctness(self):
 
+    @staticmethod
+    def _validate_zero_token_backend_support(backend: str):
+        assert backend == "aws", "Only AWS supports zero nodes configuration"
```
Assert will kill test

aleksbykov commented on sdcm/utils/common.py:
```diff
@@ -2670,5 +2670,5 @@ def skip_optional_stage(stage_names: str | list[str]) -> bool:
     if skipped_stages:
         skipped_stages_str = ', '.join(skipped_stages)
         LOGGER.warning("'%s' test stage(s) is disabled.", skipped_stages_str)
-        return False
```
done

aleksbykov commented on sdcm/cluster.py:
```diff
@@ -4300,6 +4304,59 @@ def update_db_packages(self, node_list=None, start_service=True):
                 node_list = self.nodes
             self._update_db_packages(new_scylla_bin, node_list, start_service=start_service)
 
+    @retrying(n=3, sleep_time=5)
```
implemented

aleksbykov commented:
@patjed41 @soyacz , can you review again?

soyacz commented on sdcm/nemesis.py:
```diff
@@ -253,6 +278,7 @@ def __init__(self, tester_obj, termination_event, *args, nemesis_selector=None,
         }
         self.es_publisher = NemesisElasticSearchPublisher(self.tester)
         self._init_num_deletions_factor()
+        self.set_target_node_pool(self.cluster.data_nodes)
```
more clear would be to explicitly show protected attribute `_target_node_pool`:
`self._target_node_pool = self.cluster.data_nodes`

soyacz commented on sdcm/nemesis.py:
```diff
@@ -180,6 +180,30 @@ class DefaultValue:  # pylint: disable=too-few-public-methods
     ...
 
 
+def use_data_nodes_as_target(func):
```
consider naming in the form of: `target_data_nodes`, `target_zero_nodes` and `target_all_nodes`

aleksbykov replied:
done

soyacz commented on test-cases/longevity/longevity-multi-dc-rack-aware-with-znode-in-diff_dc.yaml:
```diff
@@ -0,0 +1,25 @@
+n_db_zero_token_nodes: '0 1 1'
+
+instance_type_db: 'i4i.4xlarge'
+zero_token_instance_type_db: 'i4i.xlarge'
```
i4i.large won't be enough?

aleksbykov replied:
Switched to i4i.large. 
Also open issue https://github.com/scylladb/scylladb/issues/21321 to get answer regarding instance with ebs-only storages. 
i checked, we already can configure additional ebs storage, but for now this config will affect all nodes. This could be fixed in another PR, so we can configure data_ebs_storage separately for data and zero nodes (at least for aws via provisioner)

kbr-scylla replied:
@patjed41 made a good point on the issue. Let's just verify this experimentally.

aleksbykov commented on configurations/zerotoken_nodes/zerotoken_node_i4ilarge_1_num_single_dc.yaml:
```diff
@@ -0,0 +1,3 @@
+n_db_zero_token_nodes: 1
+zero_token_instance_type_db: 'i4i.large'
```
it looks yes: I opened issue to get clear answer https://github.com/scylladb/scylladb/issues/21321. Also in next PR it could fixed with additional sct config params related to zero nodes allowing to config the ebs storage for zero nodes separately

aleksbykov commented:
@soyacz , fixed by your comments. Can you take a look

soyacz reviewed (commented):
I took a look at risky places (possibly not all of them) and got to this:

to fix:
* `sdcm.cluster.BaseCluster.run_node_benchmarks` - should be run on data nodes
* `sdcm.cluster.BaseScyllaCluster.get_rack_nodes` - used in k8s context, replaces db node with data node.
* `longevity_test.LongevityTest.test_custom_time` - in case of `cluster_target_size` is set, it shouldn't take into account zero nodes I believe
* places where we use `adaptive_timeout` - should operate on data nodes
* `longevity_test.LongevityTest._run_validate_large_collections_in_system` and `longevity_test.LongevityTest._run_validate_large_collections_warning_in_logs` should operate on data nodes I think
* `longevity_test.LongevityTest.all_node_ips_for_stress_command` - also should operate on data nodes I think
* `longevity_tombstone_gc_test.TombstoneGcLongevityTest._run_repair_and_major_compaction` - please verify
* `sdcm.cluster.BaseScyllaCluster.start_kms_key_rotation_thread` should be adjusted to use data nodes I think
* `sdcm.nemesis.Nemesis.disrupt_load_and_stream` - verify 
*  `SstableLoadUtils.distribute_test_files_to_cluster_nodes`
* `disrupt_nodetool_refresh` in `shards_num = self.cluster.nodes[0].scylla_shards` can take zero node and return wrong value, also I don't think this disruption should upload sstables to all nodes
* `disrupt_nodetool_enospc` - this nemesis might not work with zeros, same `disrupt_end_of_quota_nemesis`
*  `check_that_sstables_are_encrypted` - this will fail
* `_switch_to_network_replication_strategy` - `nodes_by_region` take Zero nodes into account
* `disrupt_add_remove_dc` - also needs fixing in couple places
* `disrupt_add_remove_mv` - should use data nodes as targed I think (btw. will `sdcm.utils.nemesis_utils.indexes.wait_for_view_to_be_built` work?)
* `sdcm/nemesis.py:5302` - nemesis wrapper - we verify `if num_nodes_before != num_nodes_after:` this should verify if we have the same number of zero nodes and data nodes I think
* `sdcm.scan_operation_thread.FullscanOperationBase` possibly also should base on data nodes
* `sdcm.tombstone_gc_verification_thread.TombstoneGcVerificationThread` - also should work with data nodes (validates sstables)

questions:
1. Generally, shouldn't all db cql sessions be created with data nodes?
1. Shouldn't all stress threads base on data nodes too?
1. `sdcm.cluster.BaseScyllaCluster.set_seeds` - can zero node be seed?
1. `sdcm.cluster.BaseScyllaCluster.wait_for_schema_agreement` - is schema agreement check same on zero nodes?
1. `sdcm.cluster.BaseScyllaCluster.wait_all_nodes_un` uses `check_nodes_up_and_normal` which bases on nodetool status, will it work?
1. `sdcm.cluster.BaseScyllaCluster.verify_decommission` also uses nodetool status under the hood
1. `sdcm.cluster.BaseScyllaCluster.get_db_nodes_cpu_mode` - is perftune yaml always present on zero nodes?
1. `sdcm/commit_log_check_thread.py` - verify how zero nodes should be counted here, possibly the same, but I'm not sure.
1. Do we need scylla manager agent on zero nodes?
1. Do we need special treatment of zero nodes in upgrade tests? - support for upgrade 1. test can be done in followup task.

1. Does nodetool cleanup work on Zeros?
1. Won't `disrupt_validate_hh_short_downtime` fail when validating un with nodetool on all nodes
1. Won't `sdcm.cluster.BaseScyllaCluster.get_node_status_dictionary` fail if Zero node is used?
1. `sdcm.utils.compaction_ops.CompactionOps` - should work with data nodes

to check:
1. `sdcm.cluster.BaseScyllaCluster.get_node` - it is used for nosqlbench as target and other places, see if shouldn't pin to data nodes
1. `sdcm.kafka.kafka_cluster.LocalKafkaCluster` - doesn't it break kafka connector?
1. Validate SLA tests (possibly another followup task)
2. Same alternator tests (possibly another followup task)

aleksbykov commented on sdcm/sct_config.py:
```diff
@@ -2075,10 +2089,43 @@ def __init__(self):  # noqa: PLR0912, PLR0914, PLR0915
             self["mgmt_agent_backup_config"] = AgentBackupParameters(**backup_params)
+        # Validate zero token nodes
+        if self.get("use_zero_nodes"):
+            self._validate_zero_token_backend_support(backend=cluster_backend)
+            zero_nodes_num = self.get("n_db_zero_token_nodes")
+            data_nodes_num = self.get("n_db_nodes")
+            # if number of zero nodes is set for cluster setup, check correctness of settings
+            if zero_nodes_num:
+                self._validate_zero_token_backend_support(backend=cluster_backend)
```
fixed

aleksbykov commented on sdcm/sct_config.py:
```diff
@@ -2075,10 +2089,43 @@ def __init__(self):  # noqa: PLR0912, PLR0914, PLR0915
+        data_nodes_num = [data_nodes_num] if isinstance(data_nodes_num, int) else [
+            int(i) for i in str(data_nodes_num).split()]
+        total_nodes = data_nodes_num[:]
+        # temp for aws, but should work for any backend
```
removed

aleksbykov commented on sdcm/cluster.py:
```diff
@@ -145,7 +145,7 @@
 from sdcm.utils.scylla_args import ScyllaArgParser
```
yes, latest run showed that it works

soyacz replied:
so, if nodetool status works, can we rollback workaround with `check_nodes_up_and_normal_with_cql`?

aleksbykov commented on sdcm/cluster.py:
```diff
@@ -5028,6 +5028,9 @@ def backup_keyspace(self, ks):
             node.stop_scylla_server()
 
         for node in self.nodes:
+            # zero token node contains only system keyspaces
+            if node._is_zero_token_node and 'system' not in ks:
+                continue
```
fixed

aleksbykov commented on sdcm/cluster.py:
```diff
@@ -5041,6 +5044,8 @@ def restore_keyspace(self, backup_data):
             node.stop_scylla_server()
 
         for node in self.nodes:
+            if node._is_zero_token_node and 'system' not in ks:
+                continue
```
fixed

aleksbykov commented:
I took a look at risky places (possibly not all of them) and got to this:

to fix:
> * `sdcm.cluster.BaseCluster.run_node_benchmarks` - should be run on data nodes

Fixed

> * `sdcm.cluster.BaseScyllaCluster.get_rack_nodes` - used in k8s context, replaces db node with data node.

i set it as is, because k8s is not support yet. I thinks this better to fix when k8s support with zero nodes will be added
Now for K8S self.nodes == self.data_nodes

> * `longevity_test.LongevityTest.test_custom_time` - in case of `cluster_target_size` is set, it shouldn't take into account zero nodes I believe

Set to use data_nodes for counting

> * places where we use `adaptive_timeout` - should operate on data nodes. fix places where we used cluster.nodes[0]

fixed

> * `longevity_test.LongevityTest._run_validate_large_collections_in_system` and `longevity_test.LongevityTest._run_validate_large_collections_warning_in_logs` should operate on data nodes I think

Fixed. 

> * `longevity_test.LongevityTest.all_node_ips_for_stress_command` - also should operate on data nodes I think

zero nodes could be query coordinator, and driver filter them out, but set to data_nodes for safety

> * `longevity_tombstone_gc_test.TombstoneGcLongevityTest._run_repair_and_major_compaction` - please verify

switch to data_nodes

> * `sdcm.cluster.BaseScyllaCluster.start_kms_key_rotation_thread` should be adjusted to use data nodes I think

it is related to scylla config, so it is better to run on all nodes

> * `sdcm.nemesis.Nemesis.disrupt_load_and_stream` - verify 

Fixed. set to use data nodes explicitly

> *  `SstableLoadUtils.distribute_test_files_to_cluster_nodes`

It uses the passed node list, so fix above should also fix this one

> * `disrupt_nodetool_refresh` in `shards_num = self.cluster.nodes[0].scylla_shards` can take zero node and return wrong value, also I don't think this disruption should upload sstables to all nodes

fixed

> * `disrupt_nodetool_enospc` - this nemesis might not work with zeros, same `disrupt_end_of_quota_nemesis`

set data_nodes

> *  `check_that_sstables_are_encrypted` - this will fail

Set to use only data nodes. it should work. i will tirgger job to check it

> * `_switch_to_network_replication_strategy` - `nodes_by_region` take Zero nodes into account

set to use data nodes

> * `disrupt_add_remove_dc` - also needs fixing in couple places

updated with data nodes

> * `disrupt_add_remove_mv` - should use data nodes as targed I think (btw. will `sdcm.utils.nemesis_utils.indexes.wait_for_view_to_be_built` work?)

yes, it should work

> * `sdcm/nemesis.py:5302` - nemesis wrapper - we verify `if num_nodes_before != num_nodes_after:` this should verify if we have the same number of zero nodes and data nodes I think

added checks for both

> * `sdcm.scan_operation_thread.FullscanOperationBase` possibly also should base on data nodes

yep, fixed

> * `sdcm.tombstone_gc_verification_thread.TombstoneGcVerificationThread` - also should work with data nodes (validates sstables)

set to data_nodes

questions:
> 1. Generally, shouldn't all db cql sessions be created with data nodes?
no, zero node could be query coordinator and it send query request to required data node

> 1. Shouldn't all stress threads base on data nodes too?
Same as above. Dirvers will automatically ignore zero nodes even if they set in -nodes.
The only limitation is for multidc, for DC with zero nodes only should be explicit set replication factor: 0. Auto resize rf is not yet working. Issue was opened for that

> 1. `sdcm.cluster.BaseScyllaCluster.set_seeds` - can zero node be seed?
yes. 

> 1. `sdcm.cluster.BaseScyllaCluster.wait_for_schema_agreement` - is schema agreement check same on zero nodes?
yes, schema should be same on all nodes. It is supported by raft

> 1. `sdcm.cluster.BaseScyllaCluster.wait_all_nodes_un` uses `check_nodes_up_and_normal` which bases on nodetool status, will it work?
i added workaround while issue is not fixed. Status will be checked from system tables and generate same dict as nodetool status

> 1. `sdcm.cluster.BaseScyllaCluster.verify_decommission` also uses nodetool status under the hood
same


> 1. `sdcm.cluster.BaseScyllaCluster.get_db_nodes_cpu_mode` - is perftune yaml always present on zero nodes?

for now zero nodes always run on supported by scylla instances, and perftune is present. in next we could changed.

> 1. `sdcm/commit_log_check_thread.py` - verify how zero nodes should be counted here, possibly the same, but I'm not sure.

Looks same.

> 1. Do we need scylla manager agent on zero nodes?

issue opened for scylla manger zero node support. But i think yes

> 1. Do we need special treatment of zero nodes in upgrade tests? - support for upgrade 1. test can be done in followup task.

it will be next tasks. From upgrade perspective noting should be changed, and zero node upgraded as usual

> 1. Does nodetool cleanup work on Zeros?

yes, finished very fast

> 1. Won't `disrupt_validate_hh_short_downtime` fail when validating un with nodetool on all nodes

workaround should fixed it

> 1. Won't `sdcm.cluster.BaseScyllaCluster.get_node_status_dictionary` fail if Zero node is used?

workaround should help

> 1. `sdcm.utils.compaction_ops.CompactionOps` - should work with data nodes

switched

to check:
> 1. `sdcm.cluster.BaseScyllaCluster.get_node` - it is used for nosqlbench as target and other places, see if shouldn't pin to data nodes

Set to use data_nodes

> 1. `sdcm.kafka.kafka_cluster.LocalKafkaCluster` - doesn't it break kafka connector?

it shouldn't. 

> 1. Validate SLA tests (possibly another followup task)

Yes it will be another task, for now Enterprise 2024.2 is not support zero nodes yet

> 1. Same alternator tests (possibly another followup task)

Yes it will be check with another task by enabling flag for appropriate job

soyacz commented:
> 
> > * `sdcm.cluster.BaseScyllaCluster.start_kms_key_rotation_thread` should be adjusted to use data nodes I think
> 
> it is related to scylla config, so it is better to run on all nodes
> 
but inside it verifies sstables of user keyspaces - this will fail. This part needs to be adjusted.

> 
> to check:
> 
> > 1. `sdcm.cluster.BaseScyllaCluster.get_node` - it is used for nosqlbench as target and other places, see if shouldn't pin to data nodes
> 
> Set to use data_nodes

In that case, if drivers ignore zero nodes, this change is not required

> 
> > 1. `sdcm.kafka.kafka_cluster.LocalKafkaCluster` - doesn't it break kafka connector?
> 
> it shouldn't.

Worth to schedule a test for it (can be done later, and raise scylla issue if it failed)


More notes:
I didn't verify if you indeed fixed all my noted places - too hard, so make sure you pushed all changes.
see where we use `nodetool status` and adjust places use it - I might miss something.
Also you can also double check all the places where `cluster.nodes` or `self.nodes` is used - I did it, but could miss some places as it's quite mundane work.

Let us know when you finish it and is ready for merge.
Also rerun all the tests after recent updates. Include 1-2 artifact tests if don't break as this is showstopper for Scylla master and we don't want to affect that (they run on GCE, so it will verify if something required was missing there). Run also one Azure test (can be artifact too).

aleksbykov commented:
>> * `sdcm.cluster.BaseScyllaCluster.start_kms_key_rotation_thread` should be adjusted to use data nodes I think
> > 
> > 
> > it is related to scylla config, so it is better to run on all nodes
> 
> but inside it verifies sstables of user keyspaces - this will fail. This part needs to be adjusted.

Yep, i found what you mean. Fixed to use data nodes

> 
> > to check:
> > > 1. `sdcm.cluster.BaseScyllaCluster.get_node` - it is used for nosqlbench as target and other places, see if shouldn't pin to data nodes
> > 
> > 
> > Set to use data_nodes
> 
> In that case, if drivers ignore zero nodes, this change is not required

reverted.

> > > 1. `sdcm.kafka.kafka_cluster.LocalKafkaCluster` - doesn't it break kafka connector?
> > 
> > 
> > it shouldn't.
> 
> Worth to schedule a test for it (can be done later, and raise scylla issue if it failed)

Add it to plan. Can you point me to kafka job?

> 
> More notes: I didn't verify if you indeed fixed all my noted places - too hard, so make sure you pushed all changes. see where we use `nodetool status` and adjust places use it - I might miss something. Also you can also double check all the places where `cluster.nodes` or `self.nodes` is used - I did it, but could miss some places as it's quite mundane work.
> 
i ve done, but also could miss something. anyway, for now only several specific jobs will be run with enabled zero tokens, By default they will be disabled and most of the jobs should work as expected. i will run all jobs with enabling zero tokens one by one to minimize possible crashes

> Let us know when you finish it and is ready for merge. Also rerun all the tesits after recent updates. Include 1-2 artifact tests if don't break as this is showstopper for Scylla master and we don't want to affect that (they run on GCE, so it will verify if something required was missing there). Run also one Azure test (can be artifact too).

i will update job results here:

soyacz commented:
> > > > 1. `sdcm.kafka.kafka_cluster.LocalKafkaCluster` - doesn't it break kafka connector?
> > > 
> > > 
> > > it shouldn't.
> > 
> > 
> > Worth to schedule a test for it (can be done later, and raise scylla issue if it failed)
> 
> Add it to plan. Can you point me to kafka job?

https://jenkins.scylladb.com/view/master/job/scylla-master/job/kafka_connectors/ , but speak with @dimakr as recently new tests were added but I think we miss still jenkins jobs for it.

> 
> > More notes: I didn't verify if you indeed fixed all my noted places - too hard, so make sure you pushed all changes. see where we use `nodetool status` and adjust places use it - I might miss something. Also you can also double check all the places where `cluster.nodes` or `self.nodes` is used - I did it, but could miss some places as it's quite mundane work.
> 
> i ve done, but also could miss something. anyway, for now only several specific jobs will be run with enabled zero tokens, By default they will be disabled and most of the jobs should work as expected. i will run all jobs with enabling zero tokens one by one to minimize possible crashes
> 

Sure, just minimizing number of problems - we have many nemesis and better to pass as many as possible in the first run.

aleksbykov commented:
All jobs run with disabled zero nodes supports but with new nodes poool:
 .nodes = all nodes
 .data_nodes = only data nods
 .zero_nodes = only zero nodes

and new nemesis target_node_pool which is set exlicitly to use data_nodes as target 

ArtifactsJobs:
 - Passed - https://jenkins.scylladb.com/job/scylla-staging/job/abykov/job/artifacts-ubuntu-2204-test/7
 - Passed -  https://jenkins.scylladb.com/job/scylla-staging/job/abykov/job/artifacts-azure-image-test/8
 - FAILED - https://jenkins.scylladb.com/job/scylla-staging/job/abykov/job/artifacts-amazon2-test/2 - Failed with same error as in master


Invidual Nemesis, Passed (from PR point of view): Job marked as failed. But errors found in teardown. no errors in code or nemesis related to PR changes:
 - https://jenkins.scylladb.com/view/master/job/scylla-master/job/abykov-nemesis-individuals/job/longevity-5gb-1h-AbortRepairMonkey-aws-test/
 - https://jenkins.scylladb.com/view/master/job/scylla-master/job/abykov-nemesis-individuals/job/longevity-5gb-1h-AddDropColumnMonkey-aws-test/
 - https://jenkins.scylladb.com/view/master/job/scylla-master/job/abykov-nemesis-individuals/job/longevity-5gb-1h-AddRemoveDcNemesis-aws-test/
 - https://jenkins.scylladb.com/view/master/job/scylla-master/job/abykov-nemesis-individuals/job/longevity-5gb-1h-AddRemoveMvNemesis-aws-test/
 - https://jenkins.scylladb.com/view/master/job/scylla-master/job/abykov-nemesis-individuals/job/longevity-5gb-1h-ClusterRollingRestart-aws-test/

Regular jobs:
 - https://jenkins.scylladb.com/job/scylla-staging/job/abykov/job/longevity-3h-azure-test/5 (failed by spot termination) nemesis passed, no code errors
 - https://jenkins.scylladb.com/job/scylla-staging/job/abykov/job/longevity-3h-10gb-gce-test/6 (failed with cs errors on teardown) nemesis are passed, no code errors
 - https://jenkins.scylladb.com/job/scylla-staging/job/abykov/job/longevity-50gb-3days-test/5 (failed but not related to pr changes)


Jobs with zero nodes support enabled and nemesis which run with zero nodes and :
 - https://jenkins.scylladb.com/job/scylla-staging/job/abykov/job/longevity-multi-dc-rack-aware-zero-token-dc/37 - Job is passed.  Found only log errors and c-s errors
 - https://jenkins.scylladb.com/job/scylla-staging/job/abykov/job/longevity-4h-100gb-zero-nodes-larger-than-data-nodes/2 - Passed
 - https://jenkins.scylladb.com/job/scylla-staging/job/abykov/job/longevity-100gb-4h-zero-token-nodes/87/


 -

aleksbykov commented:
@soyacz can you take a look

soyacz reviewed (commented):
One comment related to fix of scylladb/scylladb#20909
otherwise,  LGTM

aleksbykov commented:
@soyacz , i removed workaround. i think PR is ready to be merged, and it have chance to run on weekend for all jobs. ( if will be something broken, to be reverted on Monday )

fruch reviewed (approved):
LGTM
